### ISSUE TITLE
mouse support

### ISSUE THREAD

**Issue by coreyoconnor** [coreyoconnor/vty#38]
Support mouse events where possible.

**Comment by jtdaugherty** [comment]
This is something I'd love to see implemented, but I don't know where I'd start if I were to implement it myself. Do you have any thoughts on this, including where to start and pitfalls?

**Comment by jtdaugherty** [comment]
This looks like a good resource:

http://www.x.org/docs/xterm/ctlseqs.pdf

**Comment by simonmichael** [comment]
I don't know the current status of this, but I just thought I'd mention that a brick+vty app in Apple's Terminal on El Capitan (at least) sees mouse-scroll events as cursor up/down, which is rather nice.

**Comment by jtdaugherty** [comment]
@coreyoconnor - I've been working on an implementation of this for the last couple of days, and it works.  But one thing I'm running into is that for terminfo-based terminals, almost no terminfo databases report any of the capabilities related to mouse events (and in the rare event that they do, they aren't useful anyway: they only report old-style capabilities and I want the extended mouse events, which are superior).  And when I run Vty applications I almost always run them in `tmux`, where Vty chooses to use the terminfo-based driver since `TERM` is not `xterm`.

The consequence of this is that my `XtermColor` implementation of this works just fine, but that isn't the one that gets chosen by Vty.  I could emit the right control sequences in the terminfo-based driver code, but I'm afraid that would not be a good idea.

Any thoughts?

**Comment by jtdaugherty** [comment]
One possibility would be to extend Vty's whitelist of terminals to include `screen` and `tmux` variants for XtermColor.  Tmux and screen both require that `TERM` be set to a variant of those.

**Comment by jtdaugherty** [comment]
This is now available in my fork of the repository on a branch at:

https://github.com/jtdaugherty/vty/tree/feature/mouse-events

**Comment by jtdaugherty** [comment]
I've also just extended this to use the `screen` terminal type check I mentioned above.  As part of that it now handles "normal mode" mouse events.  In normal mode, it is not specified which mouse button was released, so button-release events in Vty now optionally include the released button or `Nothing` if one wasn't available.

This kind of impoverishment in the API makes me think there ought to be some way to ask Vty what kind of mouse functionality can expected from the terminal on startup so that applications can degrade expectations accordingly (or refuse to run unless a better mouse event protocol mode is supported).
